[wip] updating reproschema commands to the new pydantic model

djarecka opened the pull request:
@satra - for now I updated only `validate`, but can you please take a look if this is what you had in mind?

also, I wasn't sure where the context should be, for now I added to `models`

satra commented:
looks like a reasonable approach. a few comments.

- i don't mind a temporary inclusion of the context, but ideally it should just be picked up from the release in reproschema.
- perhaps change the pydantic model file to be `model_autogen.py` and then have a model.py that simply imports and adjusts (since pydantic 2.0 has an injection mode for modification) any of the autogen classes. `validate` can then import from `model` as it does.
- reproschema2redcap and vice versa should also use these models
- we will need some helper functions probably to reduce some boilerplate of instantiating these classes (like builders).

djarecka commented:
ok, I understand that all the changes that we were doing to the pydantic model, including changes the type `langString` to `Dict[str, str]`, should be done here in `model.py`.

djarecka commented:
@satra - should we include in reproschema something like `_ldmeta` (from  dandi schema) to save that a class is `reproschema:Protocol`, etc.?
Something would be needed if we want to keep tests like [this](https://github.com/ReproNim/reproschema-py/blob/9d3e929bd73dc218cbc1e13f361b8975a2f1af3e/reproschema/models/tests/test_schema.py#L17)

satra commented:
you already have that in the meaning/mapping/schema_uri components of the linkml schema. the generator does not add that to pydantic, and that may be good to add to the metadata. i thought you or puja was going to look into how to add other metadata to the generated code on the linkml side.

more specifically, those tests are really not helpful at the moment :)

djarecka commented:
> you already have that in the meaning/mapping/schema_uri components of the linkml schema. the generator does not add that to pydantic, and that may be good to add to the metadata. i thought you or puja was going to look into how to add other metadata to the generated code on the linkml side.

Yes, indeed, there is a current PR that should add `linkml_metadata` automatically

satra commented on reproschema/utils.py:
```diff
@@ -52,3 +53,105 @@ def start_server(port=8000, path=None, tmpdir=None):
+        raise Exception(f"expected a list or dictionary, got {data_el}")
+
+
+def fixing_old_schema(data, copy_data=False):
```
should we call this `migrate_schema`, include a version check, and set `schemaVersion` to `1.0.0`?

satra commented on reproschema/models/reproschema:
```diff
@@ -0,0 +1,252 @@
+{
```
is this file needed in this repo?

djarecka replied:
I thought that first we will release `reproschema` and after this I will point to the specific version

satra commented on reproschema/validate.py:
```diff
@@ -57,13 +55,11 @@ def validate_dir(directory, shape_file, started=False, http_kwargs={}):
 
 
-def validate(shapefile, path):
+def validate(path):
```
should this also check `schemaVersion` and either ask user to migrate or ask them to use the older shacl-based validator?

djarecka replied:
but I thought that the idea was to use `fixing_old_schema` (or `migrate_schema`) for the older version. This is why I was checking entire `reproschema-library` with the new validation

satra replied:
but this function doesn't check anything, right?

djarecka replied:
it initialize the pydantic class, so will get an error if the arguments are off

satra reviewed (commented):
could we also do the things we discussed earlier in the thread (https://github.com/ReproNim/reproschema-py/pull/36#issuecomment-2043995308)?

djarecka commented:
> could we also do the things we discussed earlier in the thread ([#36 (comment)](https://github.com/ReproNim/reproschema-py/pull/36#issuecomment-2043995308))?

yes, I will work from the new redcap2reproschema scripts that @yibeichan updated

satra commented:
@djarecka - do make sure there are round robin tests: load -> save -> compare

djarecka commented:
@ibevers - check the `test_shcema` now. Yesterday it was a combination of `load_file` incorrectly treating my paths as url and that I provided wrong link to the reproschema context

djarecka commented:
> @djarecka - do make sure there are round robin tests: load -> save -> compare

yes, I just had some issues and wanted to push it to share unfinished tests with  Isaac. You can see some test now in `test_schema`.

Also, I've changed slightly the `load_file` since it was causing issues for my tests (not sure why wasn't causing issues earlier) and the logic was not clear to me. I think it tried to read my file as url. Now I just created if/else for either url or file, and use server for file. Hope I didn't miss any usecase that you wanted cover

djarecka commented:
@yibeichan - when you have a chance you can check new `redcap2reproschema`. There are still some things that want to update (e.g. moving the creation of the directory structure for activity and items to the `write_obj_jsonld` as Satra suggested), but you can let me know if you have other suggestions

yibeichan commented:
the result looks good! I didn't get errors this time

djarecka commented:
@yibeichan - could you please check if new reproschema2redcap works for you?

djarecka commented:
@yibeichan - I fixed some issues with context, but you have to create a new `b2ai-redcap2rs` with `redcap2reproschema` since context has changed. 

Also, I tried to do the test to compare csv aftter running `redcap2reproschema` and `reproschema2redcap` to the original one and I see some of the columns are empty. Perhaps some of them are left as not important, but perhaps other we should find a way to add it? One field I removed during the changes was `requiredValue`  (` row_data["required"] = ""  # response_options.get("requiredValue", "")`) since it was never a field from `ResponseOption`, but should be taken from the activity schema

yibeichan commented:
Hi, I created a new `b2ai-redcap2rs`, and this time, it worked! 

I noticed that some columns are empty. For example, branch logic, I wasn't sure whether b2ai has branch logic, so I tested it on [nimh-minimal](https://github.com/ReproNim/nimh-minimal). I git clone our nimh-minimal repo and run `reproschema2redcap` and got the following errors:
```
reproschema reproschema2redcap nimh-minimal nimh-minimal.csv
Found schema file: None
Found schema file: nimh-minimal/nimh_minimal/nimh_minimal_schema
Traceback (most recent call last):
  File "/Users/yibeichen/miniconda3/envs/tes-reproschema/bin/reproschema", line 33, in <module>
    sys.exit(load_entry_point('reproschema', 'console_scripts', 'reproschema')())
  File "/Users/yibeichen/miniconda3/envs/tes-reproschema/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/yibeichen/miniconda3/envs/tes-reproschema/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/yibeichen/miniconda3/envs/tes-reproschema/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/yibeichen/miniconda3/envs/tes-reproschema/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/yibeichen/miniconda3/envs/tes-reproschema/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/yibeichen/Desktop/test/reproschema-py/reproschema/cli.py", line 127, in reproschema2redcap
    rs2redcap(input_path_obj, output_csv_path)
  File "/Users/yibeichen/Desktop/test/reproschema-py/reproschema/reproschema2redcap.py", line 260, in main
    csv_data = get_csv_data(input_dir_path, contextfile, http_kwargs)
  File "/Users/yibeichen/Desktop/test/reproschema-py/reproschema/reproschema2redcap.py", line 179, in get_csv_data
    item_json = load_file(
  File "/Users/yibeichen/Desktop/test/reproschema-py/reproschema/jsonldutils.py", line 94, in load_file
    raise Exception(f"{path_or_url} is not a valid URL or file path")
Exception: nimh-minimal/nimh_minimal/../activities/demo/https:/raw.githubusercontent.com/ReproNim/reproschema-library/80867e36fb2c00563290486bf3f3bbeb3198f5cb/activities/NDA/items/interview_age is not a valid URL or file path
```
In nimh-minimal, we cited items from reproschema-library, see [this one](https://github.com/ReproNim/nimh-minimal/blob/main/activities/demo/demo_schema); this might need extra work for parsing.

djarecka commented:
> hello hello, I was using `redcap2reproschema` to convert HBCD redcap csv to reproschema and got the following errors, it looks like something related to input type = list. sent this csv to you on slack
> 
> ```
> Error: Error during conversion: 2 validation errors for Item
> ui.hidden
>   Extra inputs are not permitted [type=extra_forbidden, input_value=True, input_type=bool]
>     For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden
> scoringLogic
>   Extra inputs are not permitted [type=extra_forbidden, input_value=[{'variableName': 'sed_bm...bm_demo_roster_002 ))"}], input_type=list]
>     For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden
> ```

These fields are simply not parta of the reproschema model, but it is added in `redcap2reproschema`. So this is not related to pydantic version, but the logic in the code is outdated and should be updated. Perhaps `scoringLogic` is `compute`?

djarecka commented:
> In nimh-minimal, we cited items from reproschema-library, see [this one](https://github.com/ReproNim/nimh-minimal/blob/main/activities/demo/demo_schema); this might need extra work for parsing.

I fixed the code that it works for urls, however in on place `ResponseOption` is provided as url to another jsonld. This is not supported, if we want yo have it we would have to change the code. And it's not clear to me if it the model should be changed to have `url` as an option to `ResponseOption`? (cc:@satra)

satra commented:
`responseOptions` can be a url (see: https://github.com/ReproNim/reproschema-library/blob/enh/rc4/activities/GAD7/items/gad7_1#L38) and that pattern is used in many schemas.

yibeichan commented:
ah, I was thinking about this line [here](https://github.com/ReproNim/nimh-minimal/blob/ca92203510f180e2482ccc9b25eae666914eb535/activities/demo/demo_schema#L5C13-L5C148), we cited `"demo": "https://raw.githubusercontent.com/ReproNim/reproschema-library/80867e36fb2c00563290486bf3f3bbeb3198f5cb/activities/NDA/items/"`
I thought this was the one that caused the error

yibeichan commented:
> These fields are simply not parta of the reproschema model, but it is added in `redcap2reproschema`. So this is not related to pydantic version, but the logic in the code is outdated and should be updated. Perhaps `scoringLogic` is `compute`?

yes, you mean this line https://github.com/djarecka/reproschema-py/blob/86f3699fdda55802478208b5a376dde0467c1d3b/reproschema/redcap2reproschema.py#L186, right? it should be `compute` if `compute` is part of the reproschema model

djarecka commented:
@yibei - converting `nimh-minimal` should work now.

as for the `scoringLogic` perhaps you can help me tomorrow, the code doesn't look good to me, and I'm pretty sure that it doesn't do what was meant to, so we could try to fix it together.

djarecka commented:
@yibeichan  - please check how this work for your examples, as I mentioned I believe there were multiple parts that were not working as intended that I tried to fix, hopefully didn't break something else.

A few points that wasn't sure:
- it seems like it is possible that a row can have BOTH `"Branching Logic (Show field only if...)")` and `"Field Type"] == "calc"`. Which should have higher priority when deciding on `isVis`?
- `Tex Validation Min` is mapped to `valueMin` [here](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py#L23), but this is sometimes text, e.g. `[screening_arm_1][setup_lmp]`. What should be the logic here?

@satra , a few questions to you, there are some parts of the codes that are commented or not used that I'm not really sure how I would fit to the current schema. Should I completely removed them or we are missing something in the schema (in that case I would open an issue to not forget). I'm talking about:
- [matrix_group_name/count](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py#L213)
- [identifiable](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py#L269)

satra commented:
`isVis` should be based on branching logic if it says show. i believe we were using `calc` to say a field should not be shown. so if both are true for a field, we need to ensure that both calc is there in compute and isVis is dependent on branching logic. we should see a few examples where this is true.

`matrix_group_name/count` should stay (we may use that later in the `ui` component). i don't know what `identifiable` does. would need to see an example (perhaps it's in the abcd schema or in the umass schema).

djarecka commented:
> `isVis` should be based on branching logic if it says show. i believe we were using `calc` to say a field should not be shown. so if both are true for a field, we need to ensure that both calc is there in compute and isVis is dependent on branching logic. we should see a few examples where this is true.

and what about when is  "calc", i.e. `compute` in Reproschema. Previously the code was using `hidden`, but we thought that it would be now `isVis` in the `Activity` schema.

> `matrix_group_name/count` should stay (we may use that later in the `ui` component). i don't know what `identifiable` does. would need to see an example (perhaps it's in the abcd schema or in the umass schema).

ok, will keep for now

yibeichan commented:
here are some examples where both `calc` and branching logic` exist. I think we can still keep these items hidden but adding branching logic to the activity schema where those items are. basically, when both `calc` and branching logic` exist, it means we do different calculation based on different conditions, but they are still calculations.
![image](https://github.com/ReproNim/reproschema-py/assets/20402800/69359349-7495-4247-8de1-43e0f8022aed)

for `Text validation Min`, here are some examples, can we use `Text Validation Type` to help us set them up in reproschema? e.g., we set up `valueType` along with `minValue` and `maxValue`?
![image](https://github.com/ReproNim/reproschema-py/assets/20402800/77d5b01e-a75b-42b9-b569-4a13020f26e8)

satra commented:
> and what about when is "calc", i.e. `compute` in Reproschema. 

phq9 total score is a calculated item, so that activity schema is a good example for what to do with a calc only item and `isVis`.

yibeichan commented:
also, i guess because of what we mentioned above, I still got the same errors when converting HBCD
```
ui.hidden
  Extra inputs are not permitted [type=extra_forbidden, input_value=True, input_type=bool]
    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden
scoringLogic
  Extra inputs are not permitted [type=extra_forbidden, input_value=[{'variableName': 'sed_bm...bm_demo_roster_002 ))"}], input_type=list]
    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden
```

djarecka commented:
@yibeichan - did you pull the changes?

yibeichan commented:
I did. 
![image](https://github.com/ReproNim/reproschema-py/assets/20402800/81c99dba-759b-4520-96ce-e4283c42e550)

djarecka commented:
@djarecka - perhaps you are using wrong environment? Do you use the editable install? Can you double check. In the new version there is really no "scoringLogic", see [here](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py)

If you still have issues, ping me on slack, and we can quickly meet

yibeichan commented:
my bad, i didn't install the editable one. just reinstalled. it's working now!

djarecka commented:
> my bad, i didn't install the editable one. just reinstalled. it's working now!

but please look at the output

yibeichan commented:
yes, it worked, no errors. and i checked some of those with `calc` and branching logic, looks good.

yibeichan commented:
wait a second, everything is good, but can we make the schema keys in order? like put `@context` at the beginning. everything else looks in order. 
and our current context is `reproschema/ref/linkml/contexts/reproschema` I guess we'll need to change it `reproschema/master/contexts/reproschema` once we merge this PR

```
{
    "id": "sed_bm_demo_herit_002_i_01",
    "category": "reproschema:Item",
    "description": {
        "en": "@NONEOFTHEABOVE='777,999'"
    },
    "prefLabel": {
        "en": "sed_bm_demo_herit_002_i_01"
    },
    "question": {
        "en": "American Indian or Alaska Native"
    },
    "responseOptions": {
        "choices": null,
        "valueType": "xsd:string"
    },
    "ui": {
        "inputType": "radio"
    },
    "@context": "https://raw.githubusercontent.com/ReproNim/reproschema/ref/linkml/contexts/reproschema"
}
```

djarecka commented:
> wait a second, everything is good, but can we make the schema keys in order? like put `@context` at the beginning. everything else looks in order. 

this is dictionary, so really doesn't have an order, but we can create perhaps some order dictionary in `write_obj` function. I will come back to this PR later.

> and our current context is `reproschema/ref/linkml/contexts/reproschema` I guess we'll need to change it `reproschema/master/contexts/reproschema` once we merge this PR

yes, correct, this will have to be changed

djarecka commented:
>   - `Tex Validation Min` is mapped to `valueMin` [here](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py#L23), but this is sometimes text, e.g. `[screening_arm_1][setup_lmp]`. What should be the logic here?
> 

@yibeichan - can you help me with this?

yibeichan commented:
> for `Text validation Min`, here are some examples, can we use `Text Validation Type` to help us set them up in reproschema? e.g., we set up `valueType` along with `minValue` and `maxValue`? ![image](https://private-user-images.githubusercontent.com/20402800/335606875-77d5b01e-a75b-42b9-b569-4a13020f26e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczNzE0MDksIm5iZiI6MTcxNzM3MTEwOSwicGF0aCI6Ii8yMDQwMjgwMC8zMzU2MDY4NzUtNzdkNWIwMWUtYTc1Yi00MmI5LWI1NjktNGExMzAyMGYyNmU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjAyVDIzMzE0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2MWNjOGEwZmU5ZTZhZGYyZTIxMmM5NmUwODNiZjA4NjdiMTQ2M2RlNjI2MTI5ZTk1ZGIxYzdjOWU1NWY3MTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.v7J5I-zJAjy_2oVfCYy0lXSSQHpfNziwfJj9mDYaR6I)

@djarecka would this be helpful?

djarecka commented:
@yibeichan - can you check if the fields with compute makes sense. I checked the phq9 as Satra suggested and there is indeed no isVis there

yibeichan commented:
@djarecka yes, `isVis` should be `False` for compute items. 

but the newly generated compute items are not correct. there are many repetitive items, you can see a lot of `pex_bm_apa1_flag01` here. This is from `pex_bm_apa_schema` (I can't attach the file here but sent to your slack. this is generated from the HBCD csv I shared with you on slack last time)

```
{
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
        },
        {
            "jsExpression": "if(( pex_bm_apa1_suic_001 > 0 AND pex_bm_apa1_suic_001 < 777), 1, 0)",
            "variableName": "pex_bm_apa1_flag01"
```

djarecka commented:
@yibeichan - ok, will check the repetition. 

btw., I believe I also found some error with `inputType` in the code, and I started checking more, and I realized that in one of the csv you gave me there are many `Field Types`. DO you know what we should do with `checkbox` or `file`?

yibeichan commented:
@djarecka for `checkbox` we can use `multipleChoice`
for `file`, I found it in `bridge2ai` csv, where it requires pdf. we don't have equivalent fields in reproschema (we have audioOject, videoObject, image, contentUrl), maybe we can use something in schema.org, I can think about `[DigitalDocument](https://schema.org/DigitalDocument)` or `[Thing](https://schema.org/Thing)`. what do you think? @satra

djarecka commented:
@yibeichan - just to be clear, we are talking about `inputType`, and this is [the list](https://github.com/ReproNim/reproschema-ui) I'm using as a reference what is supported by the ui.
I guess `checkbox` can be `select`, but don't see `file`. 

Also not sure what are other options that you might have in other redcap cvs. I will add perhaps some exceptions, because now there are ignored

satra commented:
the demo protocol has a file upload item. perhaps check that out.

yibeichan commented:
@djarecka it's `documentUpload` https://github.com/ReproNim/demo-protocol/blob/16a54ef5eeaef7282d9ed0410cc22ff1bfef6e71/activities/Activity1/items/document_upload_item#L11

we have it in UI here: https://github.com/ReproNim/reproschema-ui/blob/05096f07b3c27e82ee9382849b073c0896448a47/src/components/Inputs/DocumentUpload/DocumentUpload.vue#L2

djarecka commented:
> @djarecka yes, `isVis` should be `False` for compute items.
for compute there is no `isVis`

> but the newly generated compute items are not correct. there are many repetitive items, you can see a lot of `pex_bm_apa1_flag01` here. This is from `pex_bm_apa_schema` (I can't attach the file here but sent to your slack. this is generated from the HBCD csv I shared with you on slack last time)

can you check now. I think the  loops in the codes was wrong and it went multiple times through rows. Now the codes runs much faster...

yibeichan commented:
yessss! it's way faster and all repetitive results have gone!! thank you!!

yibeichan commented:
@djarecka what are the remaining issues we need to fix before merging this PR?

djarecka commented:
> the demo protocol has a file upload item. perhaps check that out.

yes, I found `documentUpload`, perhaps `reproschema-ui` should be updated to give this option (@yibeichan )

what about `sql` that I see in some rows in hbcd, there are usually some "labels", e.g. "illness label". In addition that I don't know how to represent in reproschema `inputType` (I believe we don't have sql), I'm guessing that this should be hidden?? I can even see in `Field Annotation` of the csv "@HIDDEN", but in our code ["Field Annotation" is translated to "description"](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py#L17), so I'm a bit lost...

I the "Field Annotation" I can see multiple mysterious values, that I'm pretty sue are not human-readable descriptions, but have no ide what to do with them, e.g. "@HIDDEN @NOW-SERVER" or "@NONEOFTHEABOVE=5"

yibeichan commented:
> yes, I found `documentUpload`, perhaps `reproschema-ui` should be updated to give this option (@yibeichan )

yes, it's in the UI: https://github.com/ReproNim/reproschema-ui/blob/05096f07b3c27e82ee9382849b073c0896448a47/src/components/Inputs/DocumentUpload/DocumentUpload.vue#L2 (is this what you need?)

> what about `sql` that I see in some rows in hbcd, there are usually some "labels", e.g. "illness label". In addition that I don't know how to represent in reproschema `inputType` (I believe we don't have sql), I'm guessing that this should be hidden??

for `sql` we should treat it as the same we do for `calc`; it's a calculation with a different language. so we probably need to translate SQL to JS expressions or we extract the SQL and put it in compute for now but fix it later @satra what do you think? 
and we should use `isVis: false` for `sql` too

>  I can even see in `Field Annotation` of the csv "@hidden", but in our code ["Field Annotation" is translated to "description"](https://github.com/ReproNim/reproschema-py/blob/270bb9ed26bd98cec7c026f289911e52b0e6eb01/reproschema/redcap2reproschema.py#L17), so I'm a bit lost...

sorry about this, yes, it should not be translated to `description` directly. I found most rows in `Field Annotation` have `@hidden` so we should use `Field Annotation` for multiple attributes such as `isVis`, other attributes usage see below

> I the "Field Annotation" I can see multiple mysterious values, that I'm pretty sue are not human-readable descriptions, but have no ide what to do with them, e.g. "@hidden @NOW-SERVER" or "@noneoftheabove=5"

yes, for `@LANGUAGE-CURRENT-SURVEY` we are supposed to specify something language related. It's row 1876, a radio for choosing language. so it's like depending on which language a participant chose, we need to display reproschema in that language (something like `selected_language` in the UI)

row 1877, it has `@NOW` it should be date related, same as ` @NOW-SERVER`

row 1879, it has ` @IF([event-name][visit_start_complete] != 2, @DEFAULT='1', @DEFAULT='2')` we should treat it as condition or branch logic

row 1886 ` @NONEOFTHEABOVE = '4,999,777'`, should be treated as condition/branch logic too.

row 2455 `@CALCTEXT(if([pex_bm_healthv2_preg_i_illness_019_i_03] != "", [pex_bm_healthv2_preg_i_illness_019_i_03], [pex_bm_healthv2_preg_i_illness_019_i_05]))` this item should be treated as compute.

that's a lot of variations, we probably need a dictionary about how to map them. I can take a look at all HBCD and bridge2ai CSV files and find unique values of `Field Annotation` and create a dictionary for you this weekend

djarecka commented:
@yibeichan @satra 

I want to make two main points from HBCD:

- I believe neither requirements that are marked as `sql` (e.g. `SELECT distinct label from redcap_web_service_cache WHERE value in (SELECT ....`, not marked just `calc` (e.g. `$('#pex_bm_healthv2_preg_i_illness_016_i_03-tr select option:nth(1)').prop('selected',true)`) doesn't look to me as something that could be just added  to `reproschema:compute`
An example from reproschema that I found has forms like [this](https://github.com/ReproNim/reproschema/blob/3a8aa0c62df8ef76afd1561408c78ceffba95de3/docs/tutorials/translating-an-activity.md?plain=1#L187)

- for the branching logic, I found examples [here](https://github.com/ReproNim/reproschema-library/blob/9be140d0297ae0671d50466448ae67dea86d3bfd/activities/VoiceConsent/VoiceConsent_schema#L21), so the the form from "Branching Logic" in HBCD from that looks like `[sed_bm_demo_roster_003] >= 1 AND [sed_bm_demo_roster_004_i_01] = '0'` can't be directly included in `isVis`, but at least I know that it could be translated, but for the items that could be found in `Field Annotation` I really have no idea what to do (well, `@Hidden` is easy to guess), I could add to the reproschema schema, but I really do not think the `ui` will know what to do.

These issues are completely unrelated to my changes. I've been fixing multiple old issues with the code, since I'm getting newer examples, and I can keep doing this (if someone helps me with it), but perhaps this should be done in another PR? 
However, this seems like critical issue, so if I add to `compute` or `isVis` that I see in cvs files no one should expect the schema to work with `ui`.

yibeichan commented:
> * I believe neither requirements that are marked as `sql` (e.g. `SELECT distinct label from redcap_web_service_cache WHERE value in (SELECT ....`, not marked just `calc` (e.g. `$('#pex_bm_healthv2_preg_i_illness_016_i_03-tr select option:nth(1)').prop('selected',true)`) doesn't look to me as something that could be just added  to `reproschema:compute`
>   An example from reproschema that I found has forms like [this](https://github.com/ReproNim/reproschema/blob/3a8aa0c62df8ef76afd1561408c78ceffba95de3/docs/tutorials/translating-an-activity.md?plain=1#L187)

you mean SQL cannot be directly put into `jsExpression` right? yes, I agree, that's why I asked  "for sql we should treat it as the same we do for calc; it's a calculation with a different language. so we probably need to translate SQL to JS expressions or we extract the SQL and put it in compute for now but fix it later?" 

> * for the branching logic, I found examples [here](https://github.com/ReproNim/reproschema-library/blob/9be140d0297ae0671d50466448ae67dea86d3bfd/activities/VoiceConsent/VoiceConsent_schema#L21), so the the form from "Branching Logic" in HBCD from that looks like `[sed_bm_demo_roster_003] >= 1 AND [sed_bm_demo_roster_004_i_01] = '0'` can't be directly included in `isVis`, but at least I know that it could be translated. However, for the items that could be found in `Field Annotation` I really have no idea what to do (well, `@Hidden` is easy to guess), I could add `@NONEOFTHEABOVE = '4,999,777'` to the reproschema schema, but I really do not think  `reproschema-ui` will know what to do.

so among all HBCD csv files I have, `Field Annotation` has the following unique values: `{'@NOW', '@NONEOFTHEABOVE', '@READONLY', '@IF', '@NOW-SERVER', '@LANGUAGE-CURRENT-SURVEY', '@CALCDATE', '@CALCTEXT', '@HIDDEN'}`

`@NOW, @NOW-SERVER` should be something related to date, UI has something called `[Timer](https://github.com/ReproNim/reproschema-ui/blob/05096f07b3c27e82ee9382849b073c0896448a47/src/components/Timer/Timer.vue#L51)`, could be related

`@READONLY` can be treated as description 

`@IF` can be treated as branch logic

`@CALCDATE` has examples as `@CALCDATE([screening_arm_1][setup_lmp], 0, "d")` it should do some computing but yes, it's a complicated issue

`@CALCTEXT` has examples as `@CALCTEXT(if([pex_bm_healthv2_preg_i_illness_015_i_03] != "", [pex_bm_healthv2_preg_i_illness_015_i_03], [pex_bm_healthv2_preg_i_illness_015_i_05]))`, we can put it into `isVis` but reproschema may not understand it

`@NONEOFTHEABOVE` has examples as `@NONEOFTHEABOVE='12,999,777'` this can will probably need to be specified based on the response option everytime it appears. but I'm not sure what this @NONEOFTHEABOVE is used for later.

`@LANGUAGE-CURRENT-SURVEY` is related to participant's choice of language. we have language in the UI [here](https://github.com/ReproNim/reproschema-ui/blob/05096f07b3c27e82ee9382849b073c0896448a47/src/components/Landing/Landing.vue#L47) but it's for landing page

`@HIDDEN` everytime we have it, we should set `isVis: false`

> These issues are completely unrelated to my changes. I've been fixing multiple old issues with the code, since I'm getting newer examples, and I can keep doing this (if someone helps me with it), but perhaps this should be done in another PR? However, this seems like critical issue, so if I add to `compute` or `isVis` that I see in cvs files no one should expect the schema to work with `ui`.

I agree that those are complicated but critical issues. we should get this PR merged, release out, and open another PR to fix those issues

yibeichan commented:
@djarecka anything else we need to check/update in this PR? If not, I guess we can merge it and move on (to get the release out)

djarecka commented:
Don't have good internet today, but working on a couple more commits and
later today I will add them to the pr and will create issues with things
that need to be done in addition

On Mon, Jun 10, 2024, 9:49 AM Yibei Chen ***@***.***> wrote:

> @djarecka <https://github.com/djarecka> anything else we need to
> check/update in this PR? If not, I guess we can merge it and move on (to
> get the release out)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ReproNim/reproschema-py/pull/36#issuecomment-2158574300>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMV6GT2V32AFQGC53KJRSLZGW4IZAVCNFSM6AAAAABFZZZXJKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNJYGU3TIMZQGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

djarecka commented:
@yibeichan - at the end I added many command to do the script `redcap2reproschema`, please take a look if these make sense to you. 
I decided to explicitly add instruction for all elements to `INPUT_TYPE_MAP`, and `VALUE_TYPE_MAP` (previously the default behavior was to not apply any mapping, but that lead the elements that were completely ignored by reproschame) , so if something is not in the dictionaries, you will get an error, and that means that we have to decide how it should be represented in reproschema.

re. `compute` - maybe I wasn't clear on Sat, but I believe that even the elements in `calc` would not be interpreted by `reproschema-ui`, but for now I left it (and added comment, that should be also added to issues)

If that looks goo, one thing we should change for sure, is the path to contextfile

yibeichan commented:
@djarecka thanks for the work!! I tested nimh-minimal with a roundtrip: reproschema2redcap and redcap2reproschema. Everything looks good, except the branching logic.
In nimh-minimal, we have a special branching logic case. Whether participants will answer DSM5 youth or DSM5 parents depends on the interview age (an item from the previous activity). We had some problems making the ui work for this case, but @satra found a (temporary) solution for ui. So, it's reasonable that our current (general) reproschema2redcap and redcap2reproschema cannot catch this logic. 

I'm going to merge this PR now
